xla/stream_executor/command_buffer.h

```diff
@@ -172,177 +127,126 @@ class CommandBuffer {
-  absl::Status Case(DeviceMemory<int32_t> index,
-                    std::vector<Builder> branches) {
-    return Case(kDefaultExecutionScope, index, branches);
-  }
-
-  virtual absl::Status Case(ExecutionScopeId execution_scope_id,
-                            DeviceMemory<bool> index,
-                            std::vector<Builder> branches) = 0;
-
-  // Adds a conditional Case operation to default execution scope.
-  absl::Status Case(DeviceMemory<bool> index, std::vector<Builder> branches) {
-    return Case(kDefaultExecutionScope, index, branches);
-  }
-
-  // Adds a conditional operation that will execute a command buffer constructed
-  // by the `body_builder` exactly `num_iteration` times. This means the
-  // condition is known at compile time (`num_iteration` < `loop_counter`), and
-  // does not require a `cond_builder`.
-  virtual absl::Status For(ExecutionScopeId execution_scope_id,
-                           int32_t num_iteration,
-                           DeviceMemory<int32_t> loop_counter,
-                           Builder body_builder) = 0;
-
-  // Adds a conditional For operation to default execution scope.
-  absl::Status For(int32_t num_iteration, DeviceMemory<int32_t> loop_counter,
-                   Builder body_builder) {
-    return For(kDefaultExecutionScope, num_iteration, loop_counter,
-               body_builder);
-  }
-
-  // Adds a conditional operation that will execute a command buffer constructed
-  // by the `cond_builder` that must update `pred` value, and then depending on
-  // the value might execute command buffer constructed by `body_builder` and
-  // `cond_builder`. Will continue while `pred` value (which is continuously
-  // updated by `cond_builder`) is `true`.
-  //
-  // In pseudocode:
-  //
-  //   cond_builder()
-  //   while(pred):
-  //     body_builder()
-  //     cond_builder()
-  //
-  // We use execution scope builder for the condition because we have to build
-  // condition twice: (1) before the conditional node in the scope defined by
-  // `execution_scope_id` (2) inside the loop body with default execution scope.
-  virtual absl::Status While(ExecutionScopeId execution_scope_id,
-                             DeviceMemory<bool> pred,
-                             ExecutionScopeBuilder cond_builder,
-                             Builder body_builder) = 0;
-
-  // Adds a conditional While operation to default execution scope.
-  absl::Status While(DeviceMemory<bool> pred,
-                     ExecutionScopeBuilder cond_builder, Builder body_builder) {
-    return While(kDefaultExecutionScope, pred, cond_builder, body_builder);
-  }
+  // Create a new conditional handle in the underlying
+  // graph.
+  virtual absl::StatusOr<GraphConditionalHandle> CreateConditionalHandle() = 0;
+
+  // Adds a new conditional node to the graph and creates a
+  // corresponding nested command buffer.
+  virtual absl::StatusOr<ConditionalNodeResult> CreateConditionalNode(
```

Comment:
Don't use Node suffix in API, prefer simple CreateConditional, UpdateSetIfConditionKernel

Reply:
I prefer to keep the Node suffix, as API in command_buffer.h is just a shallow wrapper of cuda driver API, where cuda api has node suffix for all node creation APIs.

Reply:
the fact that it's a node is just an implementation details of CUDA (DAG with nodes), from CommandBuffer API perspective it's a set of commands with dependencies, the fact that it's a graph with nodes is not really important, and having Node in names leaks implementation details to clients

Reply:
Maybe I am viewing CommandBuffer as just a wrapper of CUDA and ROCM, and the wrapper is very shallow, so I prefer to keep consistent with CUDA and ROCM, as they both call it a node. 

The command your mentioned is actually CommandBuffercmd, from which it creates CommandBufferCmdSequence.  As we now tracking the `node handle` in CommandBufferCmd object now, calling `node` as `Command` has more confusion to me. 

So I think we should call different name for the `node` managed by CommandBuffer, and `command` in the CommandBufferCmdSequence.